hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3114,10 +3114,12 @@ private void registerEntityNameUsage(
 						subType.getEntityName(),
 						(s, existingUse) -> finalEntityNameUse.stronger( existingUse )
 				);
-				actualTableGroup.resolveTableReference(
-						null,
-						subType.getEntityPersister().getMappedTableDetails().getTableName()
-				);
+				if ( useKind == EntityNameUse.UseKind.PROJECTION ) {
+					actualTableGroup.resolveTableReference(
+							null,
+							subType.getEntityPersister().getMappedTableDetails().getTableName()
+					);
+				}
```

Comment:
I have changed this to only resolve subtype table references when the use kind is `PROJECTION` to avoid creating "useless" treats. This was causing for example in joined inheritance all subtype tables to be left-joined while none of their columns were needed in the query.

This also got rid of the error triggered just for Derby due to an [optimizer bug](https://issues.apache.org/jira/browse/DERBY-7154).

Reply:
I think the problem with doing this is that we can't determine the correct object type for the table group if we don't join all subclass tables. Not sure if my example here will reproduce the problem that I am thinking about, but let's consider the following query: `select 1 from EntitySuperclass a where treat(a as EntitySubclass).superclassProperty = 1`

This should produce a predicate like `case when a1_1.id is not null then 1 when a1_2.id is not null then 2 ... else 0 end in (2, 3) and a1_0.superclassProperty = 1`. If we don't resolve the subclass table references, we would not get a result if there is a `EntitySubSubclass` with `superclassProperty` having a value of `1`.

Note that this only is a problem for joined inheritance if there is no discriminator column, so maybe you will have to add a condition to that if e.g. `if ( useKind == EntityNameUse.UseKind.PROJECTION || isJoinedWithoutDiscriminator( persister ) ) {`

Reply:
I do not think this is an actual problem. From what I could see, `createCaseExpression` already resolves all needed table references through `registerTypeUsage(TableGroup tableGroup)`, so this should always work. I did a quick test and I believe my change here should be fine.